zaza/openstack/charm_tests/series_upgrade/tests.py

```diff
@@ -215,5 +217,166 @@ def setUpClass(cls):
+                if app_details["subordinate-to"]:
+                    continue
+                if "easyrsa" in app_details["charm"]:
+                    logging.warn(
+                        "Skipping series upgrade of easyrsa Bug #1850121")
+                    continue
+                if "etcd" in app_details["charm"]:
+                    logging.warn(
+                        "Skipping series upgrade of easyrsa Bug #1850124")
+                    continue
+                logging.warn("About to upgrade {}".format(application))
+                if "percona-cluster" in app_details["charm"]:
+                    origin = "source"
+                    pause_non_leader_primary = True
+                    pause_non_leader_subordinate = True
+                if "rabbitmq-server" in app_details["charm"]:
+                    origin = "source"
+                    pause_non_leader_primary = True
+                    pause_non_leader_subordinate = False
+                if "nova-compute" in app_details["charm"]:
+                    pause_non_leader_primary = False
+                    pause_non_leader_subordinate = False
+                if "ceph" in app_details["charm"]:
+                    origin = "source"
+                    pause_non_leader_primary = False
+                    pause_non_leader_subordinate = False
+                if "designate-bind" in app_details["charm"]:
+                    origin = None
+                if "tempest" in app_details["charm"]:
+                    origin = None
+                if "memcached" in app_details["charm"]:
+                    origin = None
+                    pause_non_leader_primary = False
+                    pause_non_leader_subordinate = False
+                if "vault" in app_details["charm"]:
+                    origin = None
+                    pause_non_leader_primary = False
+                    pause_non_leader_subordinate = True
+                    post_upgrade_functions = [
+                        ('zaza.openstack.charm_tests.vault.setup.'
+                         'mojo_unseal_by_unit')]
+                if "mongodb" in app_details["charm"]:
+                    # Mongodb needs to run series upgrade
+                    # on its secondaries first.
+                    upgrade_group.append(
+                        series_upgrade_utils
+                        .async_series_upgrade_non_leaders_first(
+                            application,
+                            from_series=self.from_series,
+                            to_series=self.to_series,
+                            completed_machines=completed_machines,
+                            post_upgrade_functions=post_upgrade_functions))
+                    continue
+
+                # The rest are likley APIs use defaults
+                pause = pause_non_leader_subordinate
+                upgrade_group.append(
+                    series_upgrade_utils.async_series_upgrade_application(
```

Comment:
Am I being stupid or is this function missing ?

Reply:
Github hid it from you because the file it's in is so big: https://github.com/openstack-charmers/zaza-openstack-tests/pull/197/files#diff-572ab3d6fed94d02e79792f13f97053fR278 ;-)